scapy/contrib/automotive/gm/gmlan_scanner.py

```diff
@@ -0,0 +1,688 @@
+
+    def _get_initial_requests(self, **kwargs):
+        # type: (Any) -> Iterable[Packet]
+        scan_range = kwargs.pop("scan_range", range(0x100))
+        rdbi_enumerator = kwargs.pop("rdbi_enumerator", None)
+        if rdbi_enumerator is None:
+            return (GMLAN() / GMLAN_WDBI(dataIdentifier=x) for x in scan_range)
+        elif isinstance(rdbi_enumerator, GMLAN_RDBIEnumerator):
+            return (GMLAN() / GMLAN_WDBI(dataIdentifier=t.resp.dataIdentifier,
+                                         dataRecord=bytes(t.resp)[2:])
+                    for t in rdbi_enumerator.filtered_results
+                    if t.resp.service != 0x7f and len(bytes(t.resp)) >= 2)
+        else:
+            raise Scapy_Exception("rdbi_enumerator has to be an instance "
+                                  "of GMLAN_RDBIEnumerator")
+
+    def _get_table_entry_y(self, tup):
+        # type: (_AutomotiveTestCaseScanResult) -> str
+        return "0x%02x: %s" % (tup[1].dataIdentifier,
+                               tup[1].sprintf("%GMLAN_WDBI.dataIdentifier%"))
+
+    def _get_table_entry_z(self, tup):
+        # type: (_AutomotiveTestCaseScanResult) -> str
+        return self._get_label(tup[2], "PR: Writeable")
+
+
+class GMLAN_WDBISelectiveEnumerator(StagedAutomotiveTestCase):
+    @staticmethod
+    def __connector_rdbi_to_wdbi(rdbi, _):
+        # type: (AutomotiveTestCaseABC, AutomotiveTestCaseABC) -> Dict[str, Any]  # noqa: E501
+        return {"rdbi_enumerator": rdbi}
+
+    def __init__(self):
+        # type: () -> None
+        super(GMLAN_WDBISelectiveEnumerator, self).__init__(
+            [GMLAN_RDBIEnumerator(), GMLAN_WDBIEnumerator()],
+            [None, self.__connector_rdbi_to_wdbi])
+
+
+class GMLAN_SAEnumerator(GMLAN_Enumerator, StateGenerator):
+    _description = "SecurityAccess supported"
+    _transition_function_args = dict()  # type: Dict[_Edge, Tuple[int, Optional[Callable[[int], int]]]]  # noqa: E501
+
+    def _get_initial_requests(self, **kwargs):
+        # type: (Any) -> Iterable[Packet]
+        return (GMLAN() / GMLAN_SA(subfunction=x) for x in range(1, 10, 2))
+
+    def _get_table_entry_y(self, tup):
+        # type: (_AutomotiveTestCaseScanResult) -> str
+        return "Subfunction %02d" % tup[1].subfunction
+
+    def _get_table_entry_z(self, tup):
+        # type: (_AutomotiveTestCaseScanResult) -> str
+        return self._get_label(tup[2], lambda r: "PR: %s" % r.securitySeed)
+
+    def pre_execute(self, socket, state, global_configuration):
+        # type: (_SocketUnion, EcuState, AutomotiveTestCaseExecutorConfiguration) -> None  # noqa: E501
+        if global_configuration.unittest:
+            return
+
+        if cast(ServiceEnumerator, self)._retry_pkt[state] is not None:
+            # this is a retry execute. Wait much longer than usual because
+            # a required time delay not expired could have been received
+            # on the previous attempt
+            time.sleep(11)
```

Comment:
Me not like :/
Isn't there an alternative to retry without waiting ?

Reply:
This is because of the GMLAN standard. Per definition, an ECU waits 10 seconds after reboot in order to allow for security access. If you ask for security access in the meantime, the ECU responds with a NegativeResponse-RequiredTimeDelayNotExpired. 

This enumerator tries to get positive responses from ECUs. If a negative response with this response code is received, this enumerator want's to retry the request. In order to ensure to get a positive response on a retry, this delay is required here.

Reply:
But, since you mentioned this code... I think the else case is not required.

Reply:
If it's in the spec, all good !